Frameworks/Foundation/NSOperation.mm

```diff
@@ -77,41 +61,88 @@ + (BOOL)automaticallyNotifiesObserversForKey:(NSString*)key {
     return NO;
 }
 
-/**
- @Status Interoperable
-*/
-+ (id)allocWithZone:(NSZone*)zone {
-    NSOperation* ret = [super allocWithZone:zone];
+- (void)_checkReady {
+    bool newReady = YES;
 
-    ret->priv = new NSOperationPriv();
+    // If cancelled, skip this logic and set _ready to YES. Otherwise check dependencies.
+    if (![self isCancelled]) {
+        _dependenciesLock.lock();
+        int count = [_dependencies count];
 
-    return ret;
+        for (int i = 0; i < count; i++) {
+            id op = [_dependencies objectAtIndex:i];
+
+            if (![op isFinished]) {
+                newReady = NO;
+            }
+        }
+
+        _dependenciesLock.unlock();
+    }
+
+    if (_ready != newReady) {
+        [self willChangeValueForKey:@"isReady"];
+        _ready = newReady;
+        [self didChangeValueForKey:@"isReady"];
+    }
+}
+
+- (void)observeValueForKeyPath:(NSString*)keyPath ofObject:(id)object change:(NSDictionary*)change context:(void*)context {
+    if ([keyPath isEqualToString:@"isFinished"]) {
+        if (object == self) {
```

Comment:
If you use context on registration for isFinished, you can avoid a string compare.